array.hpp

```diff
@@ -0,0 +1,192 @@
+#ifndef ARRAY_HPP
```

Comment:
As talked before, we can probably just use `pragma once` here. I don't think we'll be supporting any hilarious compiler that does not recognize it. If there is ever complaints about this we can just replace again with header guards (or ignore said complaints :stuck_out_tongue: )

Reply:
I think `#pragma once` is supported across gcc/clang, but is not part of the standard: https://stackoverflow.com/questions/23696115/is-pragma-once-part-of-the-c11-standard

@Rattab for this one